### PR TITLE
Persist consent to key submission (EXPOSUREAPP-3734)

### PR DESCRIPTION
**Pull request by chiljamgossow**
Copy of closed PR

Creates new SubmissionSettings class which persists if the user has given consent for the submission of keys

Refactoring of SubmissionRepository and SubmissionService to enable DI via constructor

### CODE REVIEW

**Comment by d4rken** [comment]
The comments from this PR https://github.com/corona-warn-app/cwa-app-android/pull/1610 are not all addressed in this PR, are they?

**Comment by chiljamgossow** [comment]
> The comments from this PR #1610 are not all addressed in this PR, are they?

Yes they will be

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionSettings.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    val hasGivenConsent = prefs.createFlowPreference(
+        key = "key_submission_consent",
+        defaultValue = false
+    )
+
+    @AssistedInject.Factory
+    interface Factory : InjectedSubmissionSettingsFactory
+}
+
+interface InjectedSubmissionSettingsFactory {
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionSettings.kt:27]: Why do we need assisted inject for `SubmissionSettings`?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt [hunk 2]
```diff
@@ -1,88 +1,86 @@
...
 import java.util.Date
-
-object SubmissionRepository {
-
-    private val appScope by lazy {
-        AppInjector.component.appScope
+import javax.inject.Singleton
+
+@Singleton
+class SubmissionRepository @AssistedInject constructor(
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:32]: Why do we need assisted injection?

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt [hunk 3]
```diff
@@ -1,88 +1,86 @@
...
 
     private val testResultReceivedDateFlowInternal = MutableStateFlow(Date())
     val testResultReceivedDateFlow: Flow<Date> = testResultReceivedDateFlowInternal
 
     private val deviceUIStateFlowInternal = MutableStateFlow(DeviceUIState.UNPAIRED)
     val deviceUIStateFlow: Flow<DeviceUIState> = deviceUIStateFlowInternal
 
-    private val testResultFlow = MutableStateFlow<TestResult?>(null)
+    // to be used by new submission flow screens
+    val hasGivenConsentToSubmission = submissionSettings.hasGivenConsent.flow.asLiveData()
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:62]: Please only expose flows, `LiveData` only at the UI borders `ViewModel <> LiveData <> Fragment`.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt [hunk 4]
```diff
@@ -1,88 +1,86 @@
...
-        DeviceUIState.UNPAIRED
-    }
-
-    fun updateTestResult(testResult: TestResult) {
-        this.testResultFlow.value = testResult
-
-        if (testResult == TestResult.POSITIVE) {
-            LocalData.isAllowedToSubmitDiagnosisKeys(true)
-        }
+    private val timeStamper: TimeStamper
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt:64]: Inject TimperStamper, the idea is to be able to provide a mock one in unit tests so that we have control over time.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentModule.kt [hunk 5]
```diff
@@ -15,7 +13,7 @@ abstract class HomeFragmentModule {
     @CWAViewModelKey(HomeFragmentViewModel::class)
     abstract fun homeFragment(
         factory: HomeFragmentViewModel.Factory
-    ): CWAViewModelFactory<out CWAViewModel>
+    ): HomeFragmentViewModel.HomeViewModelFactory
```
Comment (d4rken) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentModule.kt:16]: Why was this changed?

**Review by d4rken — CHANGES_REQUESTED** [review]

**Comment by d4rken** [comment]
Taking the right direction, converting this to draft though as it currently crashes when opening, likely due to the changes to injection.

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionSettings.kt (hunk 1 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/SubmissionSettings.kt]: removed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt]: removed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt (hunk 3 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt]: removed

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt (hunk 4 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/storage/SubmissionRepository.kt]: done

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentModule.kt (hunk 5 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/main/home/HomeFragmentModule.kt]: removed

**Review by d4rken — APPROVED** [review]
lgtm :+1:

**Review by AlexanderAlferov — APPROVED** [review]
LGTM
